docs: Restricted course run ADR (#4311)
openedx/course-discovery: docs/decisions/0027-restricted-course-runs.rst

27. Restricted Course Runs
============================

Status
---------
Accepted (April 2024)

Context
---------
Course Discovery was created as a central data aggregator, aggregating information from Studio, LMS, and Ecommerce. Centralizing the catalog, Discovery offers various ways
for the information to be consumed by different consumers. While the consumers consume APIs to get the information, how the APIs work internally to provide the requested information differ greatly. The majority of the APIs work by hitting the database directly. Then, there are specific catalog and search APIs that first query ElasticSearch to search object IDs and then query the database to get the complete information against provided object IDs. The ElasticSearch
APIs are primarily used by B2B/Enterprise while non-ES APIs are used by frontend B2C and other Open edX services.

> Because course discovery is an entry point and a source of truth for consumers, the different ways of fetching the information makes it challenging to restrict the fetch selectively. It is not possible to define a product available for some consumers but hidden for others. To focus on CourseRun model, as course run is the entity that contains the course content consumed by the learners, there is a **hidden** boolean field that hides the course run from most APIs if set to True. However, that flag is never set in Discovery/Publisher. Instead, it is set from Studio (catalog_visibility) and then synced to Discovery.

Reviewer: makes -> make?

27. Restricted Course Runs
============================

Status
---------
Under Review

Context
---------

> Course Discovery was created as a central data aggregator, aggregating information from Studio, LMS,and Ecommerce. Centralizing the catalog, Discovery offers various ways

Reviewer 2: remove extra comma after LMS

Author: Its Oxford Comma.

Reviewer: add space after comma?

for the information to be consumed by different consumers. While the consumers consume APIs to get the information, how the APIs work internally to provide the requested information differ greatly. The majority of the APIs hit the database directly. Then, there are specific catalog and search APIs that first query ElasticSearch for searching and then query the database to get the full set of details. The ElasticSearch

> APIs are primarily used by B2B/Enterprise while other non-ES APIs are used by frontend B2C and other Open edX services.

Reviewer 3: Is this distinction well known to external consumers of the APIs? We should make sure this flag is well documented as to which APIs it impacts

Author: We don't have data contracts to know for sure. But it is true as far as I know (for instance, enterprise-catalog uses search/all which is a major ES powered endpoint).

Because course discovery is an entry point and a source of truth for consumers, the different ways of fetching the information makes it challenging to restrict the fetch selectively. It is not possible to define a product available for some consumers but hidden for others.

Focusing solely on CourseRun model as a course run is the entity that actually houses the content to be consumed. There is a **hidden** boolean field that hides the course run from most APIs if set to True. However, that flag is never set in Discovery/Publisher. Instead, it is set from Studio (catalog_visibility) and then synced to Discovery.

Why is there a need to have course runs be available selectively? Some of the use-cases can be:

- Market special course runs at increased/decreased price in specific regions

> - Create custom presentations for B2B customers (utilizing ES APIs) without disrupting the B2C (non-ES APIs) workflow (or vice versa)

Reviewer 3: Would make this more general: "custom presentations and Small Private Online Course (SPOC)"

Decision
----------
A new model, RestrictedCourseRuns, will be added. The new model will:

- 1-1 link with CourseRun model, and

> - Specify if the run should be restricted to ElasticSearch or non-ES APIs

Reviewer 2: Also, do we plan to add the decision we made of adding this model in both draft and non-draft entries?

> - Specify if the run should be restricted to ElasticSearch or non-ES APIs

Author: sure, will update that.

This model will not override **hidden** field in any capacity. If a course run has been marked as hidden, the RestrictedCourseRun model will have no impact on the visibility of runs across APIs and objects.

Consequences
--------------

- In all the places where the course runs are fetched, whether getting course runs or aggregating based on course run information, this new model will need to be catered to.
- Publisher would need to be updated to allow a course run to be marked as restricted.
- Unlike **hidden**, this new model will not be synced with Studio. The level of control will only apply to Discovery.
- The restricted runs will be pushed to e-commerce, as these runs are still products. However, just like Studio, the e-commerce will not have any information about the course run being restricted. This falls in line with **hidden** course runs as ecommerce does not have and need that information either.
- The restricted course runs will not be exposed in APIs by default. However, both ES and non-ES APIs will have filter/query params added to include the restricted runs.

> Alternatives Considered

Reviewer 3: Did we consider having the flag as a query parameter regardless of which APIs is called (ES/not ES).  Something like "?include_restricted=true"

This would remove the mental barrier of needed to know if an API is backed by ES or not, while still allowing consumers that want to include the restricted courses to do so.

Author: Sure, makes sense.

27. Restricted Course Runs
============================

Status
---------
Under Review

Context
---------
Course Discovery was created as a central data aggregator, aggregating information from Studio, LMS,and Ecommerce. Centralizing the catalog, Discovery offers various ways

> for the information to be consumed by different consumers. While the consumers consume APIs to get the information, how the APIs work internally to provide the requested information differ greatly. The majority of the APIs hit the database directly. Then, there are specific catalog and search APIs that first query ElasticSearch for searching and then query the database to get the full set of details. The ElasticSearch

Reviewer: > query ElasticSearch for searching and then query the db

The `for searching` part sounds odd. Maybe we can strip it off?

APIs are primarily used by B2B/Enterprise while other non-ES APIs are used by frontend B2C and other Open edX services.

Because course discovery is an entry point and a source of truth for consumers, the different ways of fetching the information makes it challenging to restrict the fetch selectively. It is not possible to define a product available for some consumers but hidden for others.

> Focusing solely on CourseRun model as a course run is the entity that actually houses the content to be consumed. There is a **hidden** boolean field that hides the course run from most APIs if set to True. However, that flag is never set in Discovery/Publisher. Instead, it is set from Studio (catalog_visibility) and then synced to Discovery.

Reviewer: The start sounds odd.

Author: As in?

Reviewer: [suggest][not sure] Focusing solely on the CourseRun model, as a course run is the entity actually housing the content to be consumed, there is a hidden boolean field ...

Also, would it be better to combine this and the previous paragraph into one?

Why is there a need to have course runs be available selectively? Some of the use-cases can be:

> - Market special course runs at increased/decreased price in specific regions

Reviewer: How do you envision this being achieved?

- Create custom presentations and Small Private Online Course (SPOC) for Business-to-Business (B2B) customers without disrupting the Business-to-consumer (B2C) workflow (or vice versa)

Decision
----------
A new model, RestrictedCourseRuns, will be added. The new model will:

> - Be 1-1 DB linked with CourseRun model

Reviewer: ```suggestion
- Have a one-to-one relation with CourseRun model
```

- Use DraftManager to have both draft and non-draft entries, linked with respective CourseRun entry
- Specify if the run should be restricted to ElasticSearch or non-ES flow.

This model will not override **hidden** field in any capacity. If a course run has been marked as hidden, the RestrictedCourseRun model will have no impact on the visibility of runs across APIs and objects.

Consequences
--------------

- In all the places where the course runs are fetched, whether getting course runs or aggregating based on course run information, this new model will need to be catered to.
- Publisher would need to be updated to allow a course run to be marked as restricted.
- Unlike **hidden**, this new model will not be synced with Studio. The level of control will only apply to Discovery.

> - The restricted runs will be pushed to e-commerce, as these runs are still products. However, just like Studio, the e-commerce will not have any information about the course run being restricted. This falls in line with **hidden** course runs as ecommerce does not have and need that information either.

Reviewer: ```suggestion
- The restricted runs will be pushed to e-commerce, as these runs are still products. However, just like Studio, e-commerce will not have any information about the course run being restricted. This falls in line with **hidden** course runs as ecommerce does not have and need that information either.
```

> - The restricted runs will be pushed to e-commerce, as these runs are still products. However, just like Studio, the e-commerce will not have any information about the course run being restricted. This falls in line with **hidden** course runs as ecommerce does not have and need that information either.

Reviewer: > This falls in line with **hidden** course runs as ecommerce does not have and need that information either

This sounds odd

> - The restricted runs will be pushed to e-commerce, as these runs are products. However, e-commerce will not have any information about the course run being restricted. This behavior will be consistent with how **hidden** course runs pushed to ecommerce do not specify the runs are hidden.

Reviewer: ```suggestion
- The restricted runs will be pushed to e-commerce, as these runs are products. However, e-commerce will not have any information about the course run being restricted. This behavior will be consistent with how **hidden** course runs pushed to ecommerce do not specify whether they are hidden.
```

- The restricted course runs will not be exposed in APIs by default. However, a query param `include_restricted` will be added to APIs to provide a way to optionally fetch the restricted runs.

Alternatives Considered
-------------------------

> - Keep using hidden field for selective filtering. However, the distinction of having control over ES and non-ES APIs would be challenging.

Reviewer: > However, the distinction of having control over ES and non-ES APIs would be challenging

It is not clear what this means.

Author: Ok, will make it clear.
